### PR TITLE
Refactor lr scheduler around CosineAnnealingWarmRestarts

### PR DESCRIPTION
**Pull request by leaves-zwx**
address: https://github.com/Oneflow-Inc/OneTeam/issues/999

Provide new lr scheduler CosineAnnealingWarmRestarts ref to [CosineAnnealingWarmRestarts](https://pytorch.org/docs/master/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html#cosineannealingwarmrestarts), meanwhile add 2 new params: `decay_rate` and `restart_limit` according to [timm.cosine](https://fastai.github.io/timmdocs/SGDR#Using-cosine-scheduler-with-timm-training-script).

Refactor `get_lr` method by adding new params `base_lr` and `step`.

Refactor lazy lr conf generate method.

Refactor lazy WarmupLR. 

Fix WarmupLR states.

### CODE REVIEW

File: python/oneflow/test/modules/test_lr_scheduler.py [hunk 1]
```diff
@@ -750,5 +750,90 @@ def test(test_case):
...
 
 
+@flow.unittest.skip_unless_1n1d()
+class ChainedSchedulerTestCase(flow.unittest.TestCase):
+    def test(test_case):
+        param = flow.nn.Parameter(flow.ones(3, 4))
+        opt = flow.optim.SGD([param], lr=1)
+        s1 = flow.optim.lr_scheduler.ConstantLR(opt, factor=0.1, total_iters=3)
+        s2 = flow.optim.lr_scheduler.ExponentialLR(opt, gamma=0.9)
+        scheduler = flow.optim.lr_scheduler.ChainedScheduler([s1, s2])
```
Comment (strint) [inline python/oneflow/test/modules/test_lr_scheduler.py:760]: 这个ChainedScheduler，torch应该还没有？

File: python/oneflow/nn/optimizer/sequential_lr.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 import bisect
 from typing import Sequence, Union
 from .optimizer import Optimizer
-from .lr_scheduler import LRScheduler, _scheduler_with_step
+from .lr_scheduler import LRScheduler
 
 
 class SequentialLR(LRScheduler):
```
Comment (strint) [inline python/oneflow/nn/optimizer/sequential_lr.py:22]: 这个和ChainedScheduler是不是可以互相替换的？

File: python/oneflow/test/modules/test_lr_scheduler.py (hunk 1 above)
Comment (leaves-zwx) [reply python/oneflow/test/modules/test_lr_scheduler.py:760]: 有啊

File: python/oneflow/nn/optimizer/sequential_lr.py (hunk 2 above)
Comment (leaves-zwx) [reply python/oneflow/nn/optimizer/sequential_lr.py:22]: 作用不一样：
https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.SequentialLR.html
https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ChainedScheduler.html

**Review by chengtbf — APPROVED** [review]

**Review by MARD1NO — APPROVED** [review]

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 3]
```diff
@@ -202,6 +197,48 @@ double MultiStepLearningRate(const MultiStepConf& conf, double lr, int64_t cur_b
...
+    int64_t interval = std::pow(conf.t_mult(), epoch);
+    epoch_steps = interval * conf.t_initial();
+    step_in_epoch = step
+                    - static_cast<int64_t>(std::floor(static_cast<double>(1 - interval)
+                                                      / (1 - conf.t_mult()) * conf.t_initial()));
+  }
+  double lr = conf.eta_min();
+  if (conf.restart_limit() == 0 || (conf.restart_limit() > 0 && epoch < conf.restart_limit())) {
+    double gamma = std::pow(conf.decay_rate(), epoch);
+    const double PI = std::atan(1.0) * 4.0;
```
[the remark this answers is not in the document]
Comment (leaves-zwx) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp]: 已改

**Comment by wyg1997** [comment]
> Refactor get_lr method by adding new params base_lr and step.

这个 PR 改了之后，用户想自定义 LRScheduler 就不能兼容了：https://github.com/Oneflow-Inc/oneflow/issues/9110
